components/automate-gateway/gateway/clients.go

```diff
@@ -480,21 +456,98 @@ func (c *clientsFactory) connectionByName(name string) (*grpc.ClientConn, error)
 	return conn, nil
 }
 
-// Populate our endpoints with default targets
-func (c *ClientConfig) ConfigureDefaultEndpoints() {
+func (c *clientsFactory) Close() error {
+	var err error
+
+	for _, con := range c.connections {
+		err = con.Close()
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// DialEndpoints dials the configured endpoints. If the configuration is missing
+// a known gRPC service target and the null backend socket has been configured,
+// it will attempt to dial the null backend.
+func (c *ClientConfig) DialEndpoints(connFactory *secureconn.Factory) (ClientConnections, error) {
+	connections := ClientConnections{}
+	c.expandEndpoints()
+
+	for service, endpoint := range c.Endpoints {
+		metricsEnabled := clientMetrics[service]
+		logctx := log.WithFields(log.Fields{
+			"service": service,
+			"target":  endpoint.Target,
+			"secure":  endpoint.Secure,
+			"metrics": metricsEnabled,
+		})
+		logctx.Info("Dialing")
+
+		var conn *grpc.ClientConn
+		var err error
+
+		opts := []grpc.DialOption{}
+		if metricsEnabled {
+			opts = append(opts,
+				grpc.WithUnaryInterceptor(grpc_prometheus.UnaryClientInterceptor),
+				grpc.WithStreamInterceptor(grpc_prometheus.StreamClientInterceptor))
+		}
+
+		if endpoint.Target == c.socketPath() {
+			opts = append(opts, grpc.WithBlock(), grpc.WithTimeout(1*time.Second))
+		}
+
+		if endpoint.Secure {
```

Comment:
This is existing code so we can flag this for potential follow-up:

- I think we should probably make the _default_ "Secure" with "Insecure" as optional.
- We seem to only apply `tracing.GlobalClientInterceptor` if Insecure is set, seems like in that case we might consider just removing it.